File: prep/c/k_r/ch4/quicksort/quicksort.c
#include <stdio.h>

void qsort(int arr[], int left, int right);

int main() {
  int i;
  int nums[] = {1, 4, 3, 6, 5, 8};
  qsort(nums, 0, 5);

  for (i = 0; i < 6; i++) {
    printf("%d\n", nums[i]);
  }
}

void qsort(int arr[], int left, int right) {
  int i, last;
  void swap(int[], int, int);

  if (left >= right) {
    return;
  }

  swap(arr, left, (left + right) / 2);
  last = left;

  for (i = left + 1; i <= right; i++) {
    if (arr[i] < arr[left]) {
      swap(arr, ++last, i);
    }
  }

  swap(arr, left, last);
  qsort(arr, left, last - 1);
  qsort(arr, last + 1, right);
}

void swap(int arr[], int i, int j) {
  int temp;
  temp = arr[i];
  arr[i] = arr[j];
  arr[j] = temp;
}

File: prep/algos/notes/sort/quicksort.py
def swap(arr: list, i: int, j: int):
    temp = arr[i]
    arr[i] = arr[j]
    arr[j] = temp


def partition(arr: list, start: int, end: int):

    pivot = arr[end]
    i = start - 1

    for j in range(start, end):
        if arr[j] <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    
    arr[i + 1], arr[end] = arr[end], arr[i + 1]
    return i + 1


def quicksort(arr: list, start: int, end: int):
    if start < end:
        pivot = partition(arr, start, end)
        print(pivot)
        quicksort(arr, start, pivot - 1)
        quicksort(arr, pivot + 1, end)


nums = [1, 7, 4, 1, 10, 9, -2]
print("UNSORTED")
print(nums)

size = len(nums)
quicksort(nums, 0, size - 1)

print("SORTED")
print(nums)
